File: SmartReceipts/Model/Distance.m
//
//  Distance.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 01/05/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "Distance.h"
#import "Price.h"
#import "WBTrip.h"
#import "FMResultSet.h"
#import "DatabaseTableNames.h"
#import "NSDecimalNumber+WBNumberParse.h"
#import "NSDate+Calculations.h"

@interface Distance ()

@property (nonatomic, assign) NSUInteger objectId;

@end

@implementation Distance

- (id)initWithTrip:(WBTrip *)trip distance:(NSDecimalNumber *)distance rate:(Price *)rate location:(NSString *)location date:(NSDate *)date timeZone:(NSTimeZone *)timeZone comment:(NSString *)comment {
    self = [super init];
    if (self) {
        _trip = trip;
        _distance = distance;
        _rate = rate;
        _location = location;
        _date = date;
        _timeZone = timeZone;
        _comment = comment;
    }
    return self;
}

- (Price *)totalRate {
    NSDecimalNumber *totalValue = [self.distance decimalNumberByMultiplyingBy:self.rate.amount];
    return [Price priceWithAmount:totalValue currencyCode:self.rate.currency.code];
}

- (void)loadDataFromResultSet:(FMResultSet *)resultSet {
    [self setObjectId:(NSUInteger) [resultSet intForColumn:DistanceTable.COLUMN_ID]];
    [self setDistance:[NSDecimalNumber decimalNumberOrZero:[resultSet stringForColumn:DistanceTable.COLUMN_DISTANCE]]];
    NSString *rateString = [resultSet stringForColumn:DistanceTable.COLUMN_RATE];
    NSString *currency = [resultSet stringForColumn:DistanceTable.COLUMN_RATE_CURRENCY];
    [self setRate:[Price priceWithAmount:[NSDecimalNumber decimalNumberOrZero:rateString] currencyCode:currency]];
    [self setLocation:[resultSet stringForColumn:DistanceTable.COLUMN_LOCATION]];
    long long int dateMilliseconds = [resultSet longLongIntForColumn:DistanceTable.COLUMN_DATE];
    [self setDate:[NSDate dateWithMilliseconds:dateMilliseconds]];
    [self setTimeZone:[NSTimeZone timeZoneWithName:[resultSet stringForColumn:DistanceTable.COLUMN_TIMEZONE]]];
    [self setComment:[resultSet stringForColumn:DistanceTable.COLUMN_COMMENT]];
}

- (BOOL)isEqual:(id)other {
    if (other == self) {
        return YES;
    }
    if (!other || ![[other class] isEqual:[self class]]) {
        return NO;
    }

    Distance *otherDistance = other;
    return self.objectId == otherDistance.objectId;
}

- (NSUInteger)hash {
    return @(self.objectId).hash;
}

- (NSString *)description {
    NSMutableString *description = [NSMutableString stringWithFormat:@"<%@: ", NSStringFromClass([self class])];
    [description appendFormat:@"id: %lu", (unsigned long)self.objectId];
    [description appendString:@">"];
    return description;
}

@end

File: SmartReceipts/Model/Distance.swift
//
//  Distance.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 17/06/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Foundation
import FMDB

@objcMembers
class Distance: NSObject, NSCopying, FetchedModel {
    private var objId = 0
    var trip: WBTrip!
    var distance: NSDecimalNumber!
    var rate: Price!
    var location: String!
    var date: Date!
    var timeZone: TimeZone!
    var comment: String?
    var lastLocalModificationTime: Date!
    
    var objectId: Int {
        get { return objId }
        set { objId = newValue }
    }
    
    required init(trip: WBTrip, distance: NSDecimalNumber, rate: Price,
                  location: String, date: Date, timeZone: TimeZone, comment: String?) {
        self.trip = trip
        self.distance = distance
        self.rate = rate
        self.location = location
        self.date = date
        self.timeZone = timeZone
        self.comment = comment
    }
    
    override required init() {
        super.init()
    }
    
    override var hash: Int { get { return NSNumber(value: objId).hash } }
    
    func totalRate() -> Price {
        let totalValue = distance.multiplying(by: rate.amount)
        return Price(amount: totalValue, currencyCode: rate.currency.code)
    }
    
    override func isEqual(_ object: Any?) -> Bool {
        if let other = object as? Distance {
            if other === self {
                return true
            }
            return objId == other.objId
        } else {
            return false
        }
    }
    
    override var description: String {
        get {
            var description = "<\(NSStringFromClass(Distance.self))"
            description += "id: \(self.objId)"
            description += ">"
            return description
        }
    }
        
    func copy(with zone: NSZone? = nil) -> Any {
        let copy = Distance(trip: trip, distance: distance, rate: rate,
                        location: location, date: date, timeZone: timeZone, comment: comment)
        copy.objId = objId
        copy.lastLocalModificationTime = lastLocalModificationTime
        return copy
    }
    
    func loadData(from resultSet: FMResultSet!) {
        objId = Int(resultSet.int(forColumn: DistanceTable.Column.Id))
        distance = NSDecimalNumber(orZero: resultSet.string(forColumn: DistanceTable.Column.Distance))
        let rateString = resultSet.string(forColumn: DistanceTable.Column.Rate)
        let currency = resultSet.string(forColumn: DistanceTable.Column.RateCurrency)
        rate = Price(amount: NSDecimalNumber(orZero: rateString), currencyCode: currency!)
        location = resultSet.string(forColumn: DistanceTable.Column.Location)
        let dateMilliseconds = resultSet.unsignedLongLongInt(forColumn: DistanceTable.Column.Date)
        date = NSDate(milliseconds: Int64(dateMilliseconds))! as Date
        let tzName = resultSet.string(forColumn: DistanceTable.Column.Timezone)
        timeZone = tzName != nil ? NSTimeZone(name: tzName!)! as TimeZone : TimeZone.current
        comment = resultSet.string(forColumn: DistanceTable.Column.Comment)
        
        let time = resultSet.longLongInt(forColumn: SyncStateColumns.LastLocalModificationTime)/1000
        lastLocalModificationTime = Date(timeIntervalSince1970: TimeInterval(time))
    }

}

